omod/src/test/java/org/openmrs/module/fhir2/providers/LocationFhirResourceProviderWebTest.java

```diff
@@ -0,0 +1,180 @@
+		Location location = new Location();
+		location.setAddress(new Address().setCity(CITY));
+		when(locationService.findLocationsByCity(CITY)).thenReturn(Collections.singleton(location));
+		
+		MockHttpServletResponse response = get("/Location?address-city=" + CITY).accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
+		assertThat(readBundleResponse(response).getEntry().size(), greaterThanOrEqualTo(1));
+	}
+	
+	@Test
+	public void findLocationsByCountry_shouldReturnBundleOfLocationsWithMatchingCountry() throws Exception {
+		Location location = new Location();
+		location.setAddress(new Address().setCountry(COUNTRY));
+		when(locationService.findLocationsByCountry(COUNTRY)).thenReturn(Collections.singleton(location));
+		
+		MockHttpServletResponse response = get("/Location?address-country=" + COUNTRY).accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
+		assertThat(readBundleResponse(response).getEntry().size(), greaterThanOrEqualTo(1));
+	}
+	
+	@Test
+	public void findLocationsByPostalCode_shouldReturnBundleOfLocationsWithMatchingAddressCode() throws Exception {
+		Location location = new Location();
+		location.setAddress(new Address().setPostalCode(POSTAL_CODE));
+		when(locationService.findLocationsByPostalCode(POSTAL_CODE)).thenReturn(Collections.singleton(location));
+		
+		MockHttpServletResponse response = get("/Location?address-postalcode=" + POSTAL_CODE).accept(FhirMediaTypes.JSON)
+		        .go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
+		assertThat(readBundleResponse(response).getEntry().size(), greaterThanOrEqualTo(1));
+	}
+	
+	@Test
+	public void findLocationsByState_shouldReturnBundleOfLocationsWithMatchingAddressState() throws Exception {
+		Location location = new Location();
+		location.setAddress(new Address().setCountry(STATE));
+		when(locationService.findLocationsByState(STATE)).thenReturn(Collections.singleton(location));
+		
+		MockHttpServletResponse response = get("/Location?address-state=" + STATE).accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), equalTo(FhirMediaTypes.JSON.toString()));
+		assertThat(readBundleResponse(response).getEntry().size(), greaterThanOrEqualTo(1));
+	}
+	
+	@Test
+	public void findLocationsByTag_shouldReturnBundleOfLocationsWithMatchingTag() throws Exception {
+		Location location = new Location();
+		location.getMeta().setTag(Collections.singletonList(new Coding(FhirConstants.OPENMRS_FHIR_EXT_LOCATION_TAG,
+		        LOGIN_LOCATION_TAG_NAME, LOGIN_LOCATION_TAG_DESCRIPTION)));
+		
+		when(locationService.findLocationsByTag(any(TokenParam.class))).thenReturn(Collections.singletonList(location));
```

Comment:
Ok. Now that we know this works, we want to actually ensure that we're generating the value we expect here, so:

1. Create a `ArgumentCaptor` object like so (in the class):

  ```java
@Captor
ArgumentCaptor<TokenParam> tagCaptor;
```

2. After the `assertThat` statements you already have, add this:

  ```java
verify(locationService).findLocationsByTag(tagCaptor.capture());
```

  `verify` is from `org.mockito.Mockito`

3. Add some checks to ensure it is called with the parameters we expect:

  ```java
assertThat(codeCaptor.getValue(), notNullValue());
assertThat(codeCaptor.getValue().getSystem, equalTo(FhirConstants.OPENMRS_FHIR_EXT_LOCATION_TAG));
assertThat(codeCaptor.getValue().getValue(), equalTo(LOGIN_LOCATION_TAG_NAME));
```